chromium_src/chrome/browser/history/history_tab_helper.cc

```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/request_otr/common/buildflags/buildflags.h"
+#include "build/build_config.h"
+
+#if BUILDFLAG(ENABLE_REQUEST_OTR)
+
+#include "chrome/browser/history/history_tab_helper.h"
+
+#include "brave/components/request_otr/browser/request_otr_tab_storage.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/keyed_service/content/browser_context_keyed_service_factory.h"
+#include "components/keyed_service/core/keyed_service_factory.h"
+#include "content/public/browser/browser_context.h"
+
+namespace {
+
+bool BraveTabRequestedOffTheRecord(content::WebContents* web_contents) {
+  if (request_otr::RequestOTRTabStorage* tab_storage =
+          request_otr::RequestOTRTabStorage::FromWebContents(web_contents)) {
+    return tab_storage->RequestedOTR();
+  }
+  return false;
+}
+
+bool BraveOTRDummyFunctionAlwaysFalse() {
+  return false;
+}
+
+}  // namespace
+
+#define IsOffTheRecord                                                 \
+  IsOffTheRecord() || BraveTabRequestedOffTheRecord(web_contents()) || \
```

Comment:
this doesn't seem right, aren't you losing the `return nullptr` in this? I think we also need a test for HistoryTabHelper that covers both IsOffTheRecord and RequestOTR. Also nit: `OffTheRecord` -> `OTR` for consistency

Reply:
I don't see why it would lose the return nullptr. Upstream code:

```if (profile->IsOffTheRecord())
    return nullptr;```

patched code:

```if (profile->IsOffTheRecord() || BraveTabRequestedOffTheRecord(web_contents()) || BraveOTRDummyFunctionAlwaysFalse())
    return nullptr;```

Reply:
I did add a test though to prove that things are not being added to the history service in situations where we claim they will not being added to the history service.

Reply:
oic now why this was confusing, please use a function-like macro with no params which will eliminate the need for `BraveOTRDummyFunctionAlwaysFalse`

Reply:
tracked in https://github.com/brave/brave-browser/issues/30306